modules/src/main/java/org/archive/modules/extractor/ExtractorRobotsTxt.java

```diff
@@ -0,0 +1,115 @@
+package org.archive.modules.extractor;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.httpclient.URIException;
+import org.archive.modules.CrawlURI;
+import org.archive.modules.extractor.ContentExtractor;
+import org.archive.modules.extractor.Hop;
+import org.archive.modules.extractor.LinkContext;
+
+public class ExtractorRobotsTxt extends ContentExtractor {
+    private static final Logger LOGGER = Logger
+            .getLogger(ExtractorRobotsTxt.class.getName());
+    private static final Pattern ROBOTS_PATTERN = Pattern
+            .compile("^https?://[^/]+/robots.txt$");
+    private static final Pattern SITEMAP_PATTERN = Pattern
+            .compile("(?i)Sitemap:\\s*(.+)$");
+
+    public static final String ANNOTATION_IS_SITEMAP = "isSitemap";
+
+    @Override
+    protected boolean shouldExtract(CrawlURI uri) {
+    	boolean shouldExtract = false;
+    	if (uri.isPrerequisite()) {
+    		shouldExtract = ROBOTS_PATTERN.matcher(uri.getURI()).matches();
+            LOGGER.finest("Checked prerequisite " + uri + " GOT " + shouldExtract);
+    	}
+        return shouldExtract;
+    }
+
+    public List<String> parseRobotsTxt(InputStream input) {
+        ArrayList<String> links = new ArrayList<>();
+        BufferedReader reader = new BufferedReader(new InputStreamReader(input));
+        try {
+            String line;
+            Matcher matcher;
+            while ((line = reader.readLine()) != null) {
+                matcher = SITEMAP_PATTERN.matcher(line);
+                if (matcher.matches()) {
+                    links.add(matcher.group(1));
+                }
+            }
+        } catch (IOException e) {
+            LOGGER.warning(e.toString());
+        }
+        return links;
+    }
+
+    @Override
+    protected boolean innerExtract(CrawlURI curi) {
+        try {
+
+            // Clone the CrawlURI and change hop path and avoid queueing
+            // sitemaps as prerequisites (i.e. strip P from hop path).
+            CrawlURI curiClone = new CrawlURI(curi.getUURI(),
+                    curi.getPathFromSeed().replace("P", ""), curi.getVia(),
+                    curi.getViaContext());
+
+            // Also copy the source over:
+            curiClone.setSourceTag(curi.getSourceTag());

```

Comment:
Oops, came across a bug. You must only copy over the `SourceTag` if it's not `null`, otherwise other code breaks. This is because `getData().hasKey(x)` is used as a test instead of checking for `null` value.

```
            if (curi.getSourceTag() != null) {
                curiClone.setSourceTag(curi.getSourceTag());
            }
```